### ISSUE TITLE
lastCalledWith

### ISSUE THREAD

**Issue by appsforartists** [sinonjs/sinon#1249]
Would you guys be amenable to a PR that adds a lastCalledWith assertion?

```javascript
spy.assert.lastCalledWith(1, 2, 3)
// equivalent to
spy.lastCall.assert.calledWith(1, 2, 3)
```

Adding an assertion here would make it easier to integrate with chai via https://github.com/domenic/sinon-chai/issues/87

**Comment by mroderick** [comment]
I can't recall when I last wanted to examine a spy on more than the first/only call, so I am having a hard time imagining what a good test using this feature would look like.

I am not opposed to expanding `sinon.assert` with more assertions. However, I would rather not expand it with convenience methods to support bad practices.

Would you mind providing a more specific usage scenario to aid understanding the need for this expansion?

**Comment by appsforartists** [comment]
I just looked through the tests in my codebase: in most places, I've been using `lastCall`, but not asserting any other calls.  The one place that I've seen where I'm using `lastCall` and expecting multiple calls is testing that a spring starts on one value and comes to rest at another.

Closing for now; will reopen if I run into this again.  Thanks.

**Comment by appsforartists** [comment]
Here's another situation where I'd like to assert `lastCall`:

I'm watching a value, which gets passed to a spy whenever it changes.  When second value goes above a certain threshold, the observed value changes.  When it descends below the threshold, it changes back.  As a [marble chart](http://reactivex.io/rxjs/manual/overview.html#marble-diagrams), it looks like this:

```
input:   ---50---199---99
tested:  ---A----B-----A
```

In this example, the test would pass 50, 199, and 99 through the system and assert that `A`, `B`, and `A` were emitted at the correct times.  If I simply use `calledWith` rather than `lastCalledWith`, the correct behavior isn't tested - the `A` emitted at 50 would pass `calledWith(A)`, even if 99 didn't cause `A` to be emitted again.

In other words, these are equivalent:

```javascript
next(50)
expect(listener).to.have.been.calledWith(A)

next(199)
expect(listener).to.have.been.calledWith(B)

next(99)
expect(listener).to.have.been.calledWith(A)
```

```javascript
next(50)
expect(listener).to.have.been.calledWith(A)

next(199)
expect(listener).to.have.been.calledWith(B)

// Even though 99 isn't passed through here, A has already been passed to
// the listener, so this succeeds when it would have failed lastCalledWith
expect(listener).to.have.been.calledWith(A)
```

**Comment by mantoni** [comment]
Hm. This leads to a bigger question: Do we want to mirror all the call count related checks as assertions? Like `firstCalledWith` and `secondCalledWith` (or even `secondlyCalledWith`?). Since there is already a way to assert the arguments of the last call, I'm not sure whether we should make the API larger here. And it's not _that_ much shorter either. 🤔

**Comment by appsforartists** [comment]
I certainly won't object to adding call-count assertions, but I'd also like to avoid coupling the two requests.  There's a clear use for `lastCalledWith` (testing the latest call to a spy that will be called multiple times in a test).  

I'd posit that it's much more common to want to assert the last call than a preceding call.  Unless the code being tested synchronously calls the spy repeatedly, you can model `secondlyCalledWith` as `lastCalledWith` at the second stage of the test. 

The reason I'd like to see this added to Sinon's assertions is so that it can be [added to sinon-chai](https://github.com/domenic/sinon-chai/issues/87) to enable tests like this:

```javascript
expect(listener).to.have.been.lastCalledWith(A)
```

Which is more fluent and more obvious than 

```javascript
expect(listener.lastCall).to.have.been.calledWith(A)
```

It may not be saving many keystrokes, but it's making the API more intuitive and approachable.

**Comment by SMenigat** [comment]
May I ask what happened to this proposal? `lastCalledWith` would be a very useful assertion in many scenarios. 
Would you be open to accept a PR for this?

**Comment by fatso83** [comment]
@SMenigat Sure. The spy/stubs API is quite large anyway. Remember that a PR would need tests and updated docs as well.
